crypto/fipsmodule/sha/sha3_test.cc

```diff
@@ -67,6 +107,195 @@ class SHA3TestVector {
               Bytes(digest_.data(), digest_length));
   }
 
+  // Test SHAKE Squeeze functionality through |EVP_Digest| APIs
+  void NISTTestVectors_SHAKESqueeze(const EVP_MD *algorithm) const {
+    size_t sqd_bytes = 0, cur_test = 0, to_sq_bytes = 0;
+
+    uint32_t digest_length = out_len_ / 8;
+    std::unique_ptr<uint8_t[]> digest(new uint8_t[digest_length]);
+    bssl::ScopedEVP_MD_CTX ctx;
+
+    // Test Final XOF
+    // Assert fail when |EVP_DigestFinalXOF| is called as a streaming API
+    ASSERT_TRUE(EVP_DigestInit(ctx.get(), algorithm));
+    ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), msg_.data(), msg_.size()));
+    ASSERT_TRUE(EVP_DigestFinalXOF(ctx.get(), digest.get(), digest_length));
+    EXPECT_EQ(Bytes(digest.get(), digest_length),
+              Bytes(digest_.data(), digest_length));
+
+    ASSERT_FALSE(EVP_DigestFinalXOF(ctx.get(), digest.get(), digest_length));
+    ASSERT_FALSE(EVP_DigestSqueeze(ctx.get(), digest.get(), digest_length));
+
+    // Test the one-shot
+    // Assert success when |EVP_Digest| is called
+    OPENSSL_memset(digest.get(), 0, digest_length);
+    ASSERT_TRUE(EVP_Digest(msg_.data(), msg_.size(), digest.get(),
+                           &digest_length, algorithm, nullptr));
+    EXPECT_EQ(Bytes(digest.get(), digest_length),
+              Bytes(digest_.data(), digest_length));
+
+    // Test Final
+    // Assert fail when |EVP_DigestFinal| is called for XOF algorithms
+    OPENSSL_memset(digest.get(), 0, digest_length);
+    ASSERT_TRUE(EVP_DigestInit(ctx.get(), algorithm));
+    ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), msg_.data(), msg_.size()));
+
+    ASSERT_FALSE(EVP_DigestFinal(ctx.get(), digest.get(), &digest_length));
+    ASSERT_FALSE(EVP_DigestFinalXOF(ctx.get(), digest.get(), digest_length));
+
+    ASSERT_TRUE(EVP_DigestInit(ctx.get(), algorithm));
+    ASSERT_TRUE(EVP_DigestUpdate(ctx.get(), msg_.data(), msg_.size()));
+    ASSERT_TRUE(EVP_DigestFinalXOF(ctx.get(), digest.get(), digest_length));
+
+    EXPECT_EQ(Bytes(digest.get(), digest_length),
+              Bytes(digest_.data(), digest_length));
+
+    ASSERT_FALSE(EVP_DigestFinalXOF(ctx.get(), digest.get(), digest_length));
+
+    // Test Final XOF after Squeeze
+    // Assert fail when |EVP_DigestFinalXOF| is called after |EVP_DigestSqueeze|
+    ASSERT_TRUE(EVP_DigestInit_ex(ctx.get(), algorithm, NULL));
+    ASSERT_TRUE(EVP_DigestSqueeze(ctx.get(), digest.get(), digest_length/2));
+    ASSERT_FALSE(EVP_DigestFinalXOF(ctx.get(), digest.get() + digest_length/2,
+                                                      digest_length/2));
+
+    // Test Update after Squeeze
+    // Assert fail when |EVP_DigestUpdate| is called after |EVP_DigestSqueeze|
+    ASSERT_TRUE(EVP_DigestInit_ex(ctx.get(), algorithm, NULL));
+    ASSERT_TRUE(EVP_DigestSqueeze(ctx.get(), digest.get(), digest_length));
+    ASSERT_FALSE(EVP_DigestUpdate(ctx.get(), msg_.data(), msg_.size()));
```

Comment:
Does this sufficiently exhaust the ways that the API could be used incorrectly? Could a consumer try calling `EVP_DigestSqueeze` on an MD5?

Reply:
`EVP_DigestSqueeze` could be only called (successfully) with `EVP_MD_flags(ctx->digest) & EVP_MD_FLAG_XOF`. Other cases, such as `EVP_DigestSqueeze` after `EVP_DigestFinalXOF` and `EVP_Digest` will be restricted by `if (ctx->digest == NULL)`; `EVP_DigestSqueeze` can be called immediately after `EVP_DigestInit` or `EVP_DigestUpdate`.

Reply:
I see in [the test coverage](https://app.codecov.io/gh/aws/aws-lc/pull/2155?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=aws) that there were a few lines missing coverage from `digest.c` around this check in both `EVP_DigestFinalXOF` and `EVP_DigestSqueeze`:
```
   if ((EVP_MD_flags(ctx->digest) & EVP_MD_FLAG_XOF) == 0) {
      OPENSSL_PUT_ERROR(DIGEST, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
      return 0;
    }
```

I was thinking that extending coverage to these lines could be added fairly easily in this test?

Reply:
True. I added extra tests in the |`NISTTestVectors`|, testing the XOF specific functions when a non XOF digest is expected.